google/resource_compute_firewall.go

```diff
@@ -1,423 +1,830 @@
-				Default:      1000,
-				ValidateFunc: validation.IntBetween(0, 65535),
-			},
-
 			"allow": {
-				Type:          schema.TypeSet,
-				Optional:      true,
-				ConflictsWith: []string{"deny"},
+				Type:     schema.TypeSet,
+				Optional: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"protocol": {
 							Type:     schema.TypeString,
 							Required: true,
 						},
-
 						"ports": {
 							Type:     schema.TypeList,
 							Optional: true,
-							Elem:     &schema.Schema{Type: schema.TypeString},
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
 						},
 					},
 				},
-				Set: resourceComputeFirewallRuleHash,
+				Set:           resourceComputeFirewallRuleHash,
+				ConflictsWith: []string{"deny"},
 			},
-
 			"deny": {
-				Type:          schema.TypeSet,
-				Optional:      true,
-				ConflictsWith: []string{"allow"},
+				Type:     schema.TypeSet,
+				Optional: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"protocol": {
 							Type:     schema.TypeString,
 							Required: true,
-							ForceNew: true,
 						},
-
 						"ports": {
 							Type:     schema.TypeList,
 							Optional: true,
-							Elem:     &schema.Schema{Type: schema.TypeString},
-							ForceNew: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
 						},
 					},
 				},
-				Set: resourceComputeFirewallRuleHash,
-
-				// Unlike allow, deny can't be updated upstream
-				ForceNew: true,
+				Set:           resourceComputeFirewallRuleHash,
+				ConflictsWith: []string{"allow"},
 			},
-
 			"description": {
 				Type:     schema.TypeString,
 				Optional: true,
 			},
-
+			"destination_ranges": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Set:           schema.HashString,
+				ConflictsWith: []string{"source_ranges", "source_tags"},
+			},
 			"direction": {
 				Type:         schema.TypeString,
-				Optional:     true,
 				Computed:     true,
-				ValidateFunc: validation.StringInSlice([]string{"INGRESS", "EGRESS"}, false),
-				ForceNew:     true,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"INGRESS", "EGRESS", ""}, false),
 			},
-
 			"disabled": {
 				Type:     schema.TypeBool,
 				Optional: true,
 			},
-
-			"project": {
-				Type:     schema.TypeString,
-				Optional: true,
-				ForceNew: true,
-				Computed: true,
+			"priority": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(0, 65535),
+				Default:      1000,
 			},
-
-			"self_link": {
-				Type:     schema.TypeString,
+			"source_ranges": {
+				Type:     schema.TypeSet,
 				Computed: true,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Set: schema.HashString,
 			},
-
-			"source_ranges": {
+			"source_service_accounts": {
```

Comment:
Oh good call, I missed that. I just checked and the API still errors, so I think we should keep it so it can be plan time. If the restrictions go away in the future, our users will probably tell us :P